File: frontend/js/views/blended_debt.js
// Blended debt rate — balance-weighted average APR across debts, total
// monthly interest, and a consolidation comparison, via /calc/blended-debt.
// Editable debt rows; updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const DEFAULT_DEBTS = [
    ['Credit card', 8000, 22.99],
    ['Auto loan', 18000, 6.5],
    ['Student loan', 24000, 5.0],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 });
const esc = (s) => String(s).replace(/[&<>"]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;' }[c]));

function debtRow(name = '', bal = 0, apr = 0) {
    return `<tr class="bd-row">
        <td><input type="text" class="bd-name" value="${esc(name)}" placeholder="${esc(t('view.bld.ph.name'))}"></td>
        <td><input type="number" step="0.01" min="0" class="bd-bal" value="${bal}"></td>
        <td><input type="number" step="0.01" min="0" class="bd-apr" value="${apr}"></td>
        <td><button type="button" class="btn btn-secondary bd-del">✕</button></td>
    </tr>`;
}

export async function renderBlendedDebt(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.bld.h1.title">// BLENDED DEBT RATE</span></h1>
        <p class="muted small" data-i18n="view.bld.hint.intro">
            With several balances at different rates, the rate that matters is the
            balance-weighted average — a big balance at a middling rate costs more than a tiny
            one at a brutal rate. Enter your debts to get the blended APR and total monthly
            interest, then a consolidation rate to see if rolling them into one loan saves
            money. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.bld.h2.inputs">Your debts</h2>
            <table class="data-table" id="bd-debts">
                <thead><tr>
                    <th data-i18n="view.bld.col.name">Debt</th>
                    <th data-i18n="view.bld.col.balance">Balance</th>
                    <th data-i18n="view.bld.col.apr">APR %</th>
                    <th></th>
                </tr></thead>
                <tbody>${DEFAULT_DEBTS.map((d) => debtRow(...d)).join('')}</tbody>
            </table>
            <p>
                <button type="button" class="btn btn-secondary" id="bd-add" data-i18n="view.bld.btn.add">+ Add debt</button>
            </p>
            <form id="bd-form" class="inline-form">
                <label><span data-i18n="view.bld.label.consol">Consolidation APR (%, 0 = skip)</span>
                    <input type="number" step="0.01" min="0" name="consolidation_apr_pct" value="9"></label>
            </form>
        </div>
        <div id="bd-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const body = mount.querySelector('#bd-debts tbody');
    const generate = debounce(async () => {
        const debts = [...mount.querySelectorAll('#bd-debts .bd-row')].map((row) => ({
            name: row.querySelector('.bd-name').value.trim(),
            balance_usd: Number(row.querySelector('.bd-bal').value) || 0,
            apr_pct: Number(row.querySelector('.bd-apr').value) || 0,
        }));
        const consolidation_apr_pct = Number(mount.querySelector('[name="consolidation_apr_pct"]').value) || 0;
        try {
            const r = await api.calcBlendedDebt({ debts, consolidation_apr_pct });
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.bld.toast.error'), { level: 'error' });
        }
    }, 250);

    mount.querySelector('#bd-add').addEventListener('click', () => {
        body.insertAdjacentHTML('beforeend', debtRow());
        applyUiI18n(body.lastElementChild);
        generate();
    });
    body.addEventListener('click', (e) => {
        if (e.target.classList.contains('bd-del')) {
            if (body.querySelectorAll('.bd-row').length > 1) { e.target.closest('tr').remove(); generate(); }
        }
    });
    body.addEventListener('input', generate);
    mount.querySelector('#bd-form').addEventListener('input', generate);
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#bd-result');
    const worth = r.consolidation_worth_it;
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.bld.h2.result">The blend</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.bld.card.blended">Blended APR</div>
                    <div class="value">${Number(r.blended_apr_pct).toFixed(2)}%</div></div>
                <div class="card"><div class="label" data-i18n="view.bld.card.balance">Total balance</div>
                    <div class="value">${money(r.total_balance_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.bld.card.interest">Monthly interest</div>
                    <div class="value neg">${money(r.total_monthly_interest_usd)}</div></div>
                <div class="card ${worth ? 'pos' : ''}"><div class="label" data-i18n="view.bld.card.consol">Consolidation</div>
                    <div class="value ${worth ? 'pos' : 'neg'}">${worth ? t('view.bld.worth') : t('view.bld.notworth')}</div></div>
            </div>
            <table class="data-table">
                <thead><tr><th data-i18n="view.bld.col.line">Line</th><th data-i18n="view.bld.col.amount">Monthly</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.bld.row.current">Current monthly interest</td><td>${money(r.total_monthly_interest_usd)}</td></tr>
                    <tr><td data-i18n="view.bld.row.consol">Consolidated monthly interest</td><td>${money(r.consolidation_monthly_interest_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.bld.row.savings">Monthly savings</td>
                        <td class="${r.monthly_savings_usd >= 0 ? 'pos' : 'neg'}">${money(r.monthly_savings_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/blended_debt.rs
//! Blended debt rate — the balance-weighted average APR across debts.
//!
//! With several balances at different rates, the single rate that matters is
//! the **balance-weighted** average — a big balance at a middling rate costs
//! more than a tiny balance at a brutal one:
//!
//!   * blended APR = Σ(balanceᵢ × APRᵢ) / Σ(balanceᵢ)
//!   * monthly interest = Σ(balanceᵢ × APRᵢ / 12)
//!
//! Compare the blended rate to a consolidation-loan rate: if the loan rate
//! is below the blended APR, consolidating cuts the monthly interest.
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct Debt {
    #[serde(default)]
    pub name: String,
    pub balance_usd: f64,
    pub apr_pct: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct BlendedDebtInput {
    pub debts: Vec<Debt>,
    /// Optional consolidation-loan APR to compare against (0 ⇒ skip).
    #[serde(default)]
    pub consolidation_apr_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct BlendedDebtResult {
    pub total_balance_usd: f64,
    pub blended_apr_pct: f64,
    pub total_monthly_interest_usd: f64,
    /// Monthly interest if all balances moved to the consolidation rate.
    pub consolidation_monthly_interest_usd: f64,
    /// Current monthly interest − consolidation monthly interest.
    pub monthly_savings_usd: f64,
    /// True when the consolidation rate beats the blended APR.
    pub consolidation_worth_it: bool,
}

pub fn analyze(i: &BlendedDebtInput) -> BlendedDebtResult {
    let total_balance: f64 = i.debts.iter().map(|d| d.balance_usd.max(0.0)).sum();
    let weighted: f64 = i.debts.iter().map(|d| d.balance_usd.max(0.0) * d.apr_pct).sum();

    let blended_apr = if total_balance > 0.0 { weighted / total_balance } else { 0.0 };
    let monthly_interest = total_balance * blended_apr / 100.0 / 12.0;

    let consol_monthly = total_balance * i.consolidation_apr_pct / 100.0 / 12.0;
    // Only treat consolidation as beneficial when a positive rate was given.
    let consol_given = i.consolidation_apr_pct > 0.0;
    let savings = monthly_interest - consol_monthly;
    let worth_it = consol_given && i.consolidation_apr_pct < blended_apr;

    BlendedDebtResult {
        total_balance_usd: total_balance,
        blended_apr_pct: blended_apr,
        total_monthly_interest_usd: monthly_interest,
        consolidation_monthly_interest_usd: consol_monthly,
        monthly_savings_usd: savings,
        consolidation_worth_it: worth_it,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn debt(b: f64, apr: f64) -> Debt {
        Debt { name: String::new(), balance_usd: b, apr_pct: apr }
    }

    #[test]
    fn blended_of_two_debts() {
        // 10k@20% + 30k@8% → (10k×20 + 30k×8)/40k = (200k+240k)/40k = 11%.
        let r = analyze(&BlendedDebtInput {
            debts: vec![debt(10_000.0, 20.0), debt(30_000.0, 8.0)],
            consolidation_apr_pct: 0.0,
        });
        assert!((r.blended_apr_pct - 11.0).abs() < 1e-9);
    }

    #[test]
    fn total_balance_sums() {
        let r = analyze(&BlendedDebtInput {
            debts: vec![debt(10_000.0, 20.0), debt(30_000.0, 8.0)],
            consolidation_apr_pct: 0.0,
        });
        assert!((r.total_balance_usd - 40_000.0).abs() < 1e-9);
    }

    #[test]
    fn monthly_interest_at_blended() {
        // 40k × 11% / 12 = 366.67.
        let r = analyze(&BlendedDebtInput {
            debts: vec![debt(10_000.0, 20.0), debt(30_000.0, 8.0)],
            consolidation_apr_pct: 0.0,
        });
        assert!((r.total_monthly_interest_usd - 40_000.0 * 0.11 / 12.0).abs() < 1e-6);
    }

    #[test]
    fn single_debt_blended_equals_its_apr() {
        let r = analyze(&BlendedDebtInput { debts: vec![debt(5_000.0, 17.99)], consolidation_apr_pct: 0.0 });
        assert!((r.blended_apr_pct - 17.99).abs() < 1e-9);
    }

    #[test]
    fn bigger_balance_dominates_the_blend() {
        // Tiny brutal balance barely moves the blend.
        let r = analyze(&BlendedDebtInput {
            debts: vec![debt(100.0, 30.0), debt(100_000.0, 5.0)],
            consolidation_apr_pct: 0.0,
        });
        assert!(r.blended_apr_pct < 5.1);
    }

    #[test]
    fn consolidation_saves_when_rate_below_blended() {
        // Blended 11%; consolidate at 9% → worth it, positive savings.
        let r = analyze(&BlendedDebtInput {
            debts: vec![debt(10_000.0, 20.0), debt(30_000.0, 8.0)],
            consolidation_apr_pct: 9.0,
        });
        assert!(r.consolidation_worth_it);
        assert!(r.monthly_savings_usd > 0.0);
    }

    #[test]
    fn consolidation_not_worth_it_above_blended() {
        let r = analyze(&BlendedDebtInput {
            debts: vec![debt(10_000.0, 20.0), debt(30_000.0, 8.0)],
            consolidation_apr_pct: 15.0,
        });
        assert!(!r.consolidation_worth_it);
        assert!(r.monthly_savings_usd < 0.0);
    }

    #[test]
    fn empty_or_zero_balance_guards() {
        let r = analyze(&BlendedDebtInput { debts: vec![], consolidation_apr_pct: 9.0 });
        assert!(r.total_balance_usd.abs() < 1e-9);
        assert!(r.blended_apr_pct.abs() < 1e-9);
        assert!(!r.consolidation_worth_it); // nothing to consolidate
    }
}
